param/parameterized.py

```diff
@@ -2049,43 +2276,54 @@ def watch(self_, fn, parameter_names, what='value', onlychanged=True, queued=Fal
 
         See help(Watcher) and help(Event) for the contents of those objects.
         """
+        if precedence < 0:
+            raise ValueError("User-defined watch callbacks must declare "
+                             "a positive precedence. Negative precedences "
+                             "are reserved for internal Watchers.")
+        return self_._internal_watch(fn, parameter_names, what, onlychanged, queued, precedence)
+
+    def _internal_watch(self_, fn, parameter_names, what='value', onlychanged=True, queued=False, precedence=-1):
         parameter_names = tuple(parameter_names) if isinstance(parameter_names, list) else (parameter_names,)
         watcher = Watcher(inst=self_.self, cls=self_.cls, fn=fn, mode='args',
                           onlychanged=onlychanged, parameter_names=parameter_names,
-                          what=what, queued=queued)
+                          what=what, queued=queued, precedence=precedence)
         self_._watch('append', watcher, what)
         return watcher
 
-    def unwatch(self_,watcher):
+    def unwatch(self_, watcher):
         """
         Remove the given Watcher object (from `watch` or `watch_values`) from this object's list.
         """
         try:
-            self_._watch('remove',watcher)
-        except:
+            self_._watch('remove', watcher, what=watcher.what)
+        except Exception:
             self_.warning('No such watcher {watcher} to remove.'.format(watcher=watcher))
 
-
-    def watch_values(self_, fn, parameter_names, what='value', onlychanged=True, queued=False):
+    def watch_values(self_, fn, parameter_names, what='value', onlychanged=True, queued=False, precedence=0):
         """
         Easier-to-use version of `watch` specific to watching for changes in parameter values.
 
         Only allows `what` to be 'value', and invokes the callback `fn` using keyword
         arguments <param_name>=<new_value> rather than with a list of Event objects.
         """
-        assert(what=='value')
-        parameter_names = tuple(parameter_names) if isinstance(parameter_names, list) else (parameter_names,)
+        if precedence < 0:
+            raise ValueError("User-defined watch callbacks must declare "
+                             "a positive precedence. Negative precedences "
+                             "are reserved for internal Watchers.")
+        assert what == 'value'
+        if isinstance(parameter_names, list):
+            parameter_names = tuple(parameter_names)
+        else:
+            parameter_names = (parameter_names,)
         watcher = Watcher(inst=self_.self, cls=self_.cls, fn=fn,
                           mode='kwargs', onlychanged=onlychanged,
                           parameter_names=parameter_names, what=what,
-                          queued=queued)
+                          queued=queued, precedence=0)
```

Comment:
Looking at the diff, precedence is checked (make sure it is > 0) but then isn't used and hard coded to zero?

Reply:
Good catch!